### ISSUE TITLE
M-x slime [No match]

### ISSUE THREAD

**Issue by 7stud** [slime/slime#808]
I installed `Emacs 29.2` and `sbcl 2.4.0` on `macOS 12.51`.

Then I installed `slime` using the command `M-x package-install<Ret>slime<Ret>`.  I can see that slime is installed:

<img width="431" alt="Screen Shot 2024-02-03 at 12 34 44 PM" src="https://github.com/slime/slime/assets/891276/0c643a2f-0d0e-4600-9c70-4439ea4ff429">

Now, I'm trying to get slime to work following [these instructions][3]:

> It is simple as:
> 
> 1. Open emacs
> 2. Open your lisp file with fibonacci function
> 3. Issue M-x slime
> 4. Place you cursor over fibonacci function and press C-c C-c to evaluate/compile it in Slime.
> 5. switch to slime window and call (fibonacci 10)
> 
> Screenshot example with hello-world function:

In emacs, I opened a hello world `.lisp` file with `C-x C-f`. However, when I type `M-x slime<Ret>`, I get `[No match]`.  Same for `M-x slime-mode<Ret>`.

Here is my `~/.emacs` file:

    (custom-set-variables
     ;; custom-set-variables was added by Custom.
     ;; If you edit it by hand, you could mess it up, so be careful.
     ;; Your init file should contain only one such instance.
     ;; If there is more than one, they won't work right.
     '(package-selected-packages '(slime paredit)))
    (custom-set-faces
     ;; custom-set-faces was added by Custom.
     ;; If you edit it by hand, you could mess it up, so be careful.
     ;; Your init file should contain only one such instance.
     ;; If there is more than one, they won't work right.
     )
    
    ;;;;;;;;;;;;;;;;;;;;;;;
    
    ;; I added the following according to slime install directions:
    
    (setq inferior-lisp-program "sbcl")

`sbcl` is in my path:

    % which sbcl
    /opt/local/bin/sbcl
    
    % echo $PATH
    ...:/Users/7stud/bin/:/opt/local/bin:....

I tried altering the .emacs file to use the full path to `sbcl`:

    (setq inferior-lisp-program "/opt/local/bin/sbcl")

but I still get `M-x slime [No match]`.  I've been quitting emacs and relaunching it after I make changes to the `.emacs` file.

Here is my `~/.emacs.d/elpa` directory (which I haven't touched):

    % ls
    archives               macrostep-0.9.2.signed slime-2.29.1
    gnupg                  paredit-26             slime-2.29.1.signed
    macrostep-0.9.2        paredit-26.signed

There were a bunch of warnings when I installed slime

**Comment by stassats** [comment]
2.29.1 is the latest version of Slime. I have never used melpa, but I've seen a lot of confusion from those who use it.

**Comment by pedro-w** [comment]
@stassats thanks for the information. When I install packages (with MELPA enabled) I have a choice of
```
 20240125.1336 (melpa), 2.29.1 (nongnu).
```
The nongnu version has a date of 2024-Jan-24 ([reference](https://elpa.nongnu.org/nongnu/slime.html))
and the MELPA version also claims to be 2.29.1
so they ought to be very similar, however it's still true that the MELPA one works for me but the nongnu one gives the [no match] error as reported by @7stud.
I can't see anything in messages to indicate why the nongnu one is not loaded.

**Comment by stassats** [comment]
The elpa/melpa/whatever packages are maintained by someone (something) else. So I have no idea what they do.

**Comment by pedro-w** [comment]
> I have never used melpa, but I've seen a lot of confusion from those who use it.

Understood, but literally item 1 in your quick setup instructions is "Set up the MELPA repository"

As far as I can see, the nongnu ELPA version seems to have inexplicably omitted your [`slime-autoloads.el`](https://github.com/slime/slime/blob/master/slime-autoloads.el) from their tarball and it gets replaced with an autogenerated version that doesn't work.

I see there was some sort of beef with ELPA recorded [here](https://git.savannah.gnu.org/cgit/emacs/nongnu.git/tree/elpa-packages)
```
 ;; FIXME: Upstream has accidentally diverged, see
 ;; https://github.com/slime/slime/pull/648
 (slime			:url "https://github.com/slime/slime"
  ;; According to `git log -- NEWS`, seems to be the closest to "maintainer".
  :maintainer "Stas Boukarev <stassats@gmail.com>"
  ;; FIXME: Apparently upstream is not cooperative enough :-(
  ;; https://github.com/slime/slime/pull/648
  :merge t
  :doc "doc/slime.texi")
```
and in the linked PR #648

Do you want me to try and contact that someone or something at ELPA because it would be good to have this working "out of the box"? I'm happy to give it a go!

**Comment by stassats** [comment]
Part of the reason I don't understand or trust ELPA, why did they include code from some third party branch? And then can't remove it. I don't really want to be fixing their mistakes.

**Comment by 7stud** [comment]
> You probably have figured this out yourself,

No.  I just got lucky when I decided to try the .emacs file from emacs4cl.  I could see some differences in `~.emacs.d/`, but I didn't know the significance of them.

**Comment by pedro-w** [comment]
I'm coming to the conclusion that this isn't the right place to complain; both MELPA and nongnu ELPA take the outputs of this project and package them in their own ways. MELPA updates theirs every time there's a new commit and ELPA take the github releases. At the moment, MELPA works and ELPA doesn't because they're modifying the github release tarball. (If I'm wrong about that, please someone correct me) 

So, if you want to get going today, use MELPA. In the meantime I will try and contact whoever is behind the ELPA and see if I can figure out why the difference and if it can be brought into sync, because I'd like slime to be working without this hassle! 

For reference, my `.emacs` looks like
```
(custom-set-variables
 ;; custom-set-variables was added by Custom.
 ;; If you edit it by hand, you could mess it up, so be careful.
 ;; Your init file should contain only one such instance.
 ;; If there is more than one, they won't work right.
 '(inferior-lisp-program "ecl" t)
 '(inhibit-startup-screen t)
 '(package-selected-packages '(slime paredit tj3-mode)))
(custom-set-faces
 ;; custom-set-faces was added by Custom.
 ;; If you edit it by hand, you could mess it up, so be careful.
 ;; Your init file should contain only one such instance.
 ;; If there is more than one, they won't work right.
 )

(require 'package)
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/") t)
(setq inferior-lisp-program "ecl")
(package-initialize)
```

**Comment by stassats** [comment]
Is the "Source code(tar.gz)" from the release page really unusable? I, naturally, haven't tried using it.

**Comment by pedro-w** [comment]
I haven't tried either (beyond looking at its contents) but I believe it *would* be usable - the one that isn't is nongnu ELPA because  they have removed `slime-autoloads.el`. During installation it is replaced with an autogenerated one which doesn't work because slime doesn't use the magic comments (`;;;###autoload`) which would allow emacs to detect which functions should be made into autoloaders.  

In other words, normally the function `slime` would be registered by code in `slime-autoloads.el` and typing `M-x slime` would cause the rest of the package to be loaded. However the 'broken' `slime-autoloads.el` doesn't register anything, so emacs doesn't know anything about function `slime`, hence `M-x slime` gives the "[no completions]" error.

If you install the ELPA package and cat `~/.emacs.d/elpa/slime-2.29.1/slime-autoloads.el` you'll see it's different from the one in this repo.

I think that's correct - let me check it before we get any further though!

**Comment by stassats** [comment]
A pull request with magic would be welcome.

**Comment by pedro-w** [comment]
I've asked a question on the emacs mailing list (hopefully the right place) as I need a bit of guidance on fixing this

**Comment by phikal** [comment]
> At the moment, MELPA works and ELPA doesn't because they're modifying the github release tarball. (If I'm wrong about that, please someone correct me)

Note that NonGNU ELPA doesn't modify any tarball, we use the last commit that bumps the version tag and use that as the basis to prepare a .tar package that is hosted by the archive.  As part of that process, the ELPA build server generates an `-autoloads.el` and `-pkg.el` file, which in this case get overwritten.

**Comment by luismbo** [comment]
@phikal slime-autoloads.el was removed from emacs/nongnu.git in this commit: https://git.savannah.gnu.org/gitweb/?p=emacs/nongnu.git;a=commit;h=6e0d67b9b7ae0fc31902609ab7525826b292c384

**Comment by phikal** [comment]
I know, that is because they don't have to be tracked in the repository.

**Comment by 7stud** [comment]
> So, if you want to get going today, use MELPA.

On stackoverflow, Ehvince asked me to try adding `(require 'slime)` to my `.emacs` file, and that got rid of the `M-x slime [No match]` proglem.   In a buffer containing a lisp program, I did `M-x slime`, and another window opened with the slime REPL.  Both `C-c C-c` with the cursor on a function name,  and `C-x C-e` with the cursor just after the closing paren of a function worked: I could switch to the slime REPL and execute the functions.
